mlflow/sklearn/__init__.py

```diff
@@ -455,6 +461,271 @@ def load_model(model_uri):
+
+        def clean_id(id_):
+            _get_autolog_training_status().pred_result_id_to_dataset_name_and_model_id \
+                .pop(id_, None)
+
+        # When the `predict_result` object being GCed, its ID may be reused, so register a finalizer
+        # to clear the ID from the dict for preventing wrong ID mapping.
+        weakref.finalize(predict_result, clean_id, prediction_result_id)
+
+        return predict_result
+
+    def register_metric_api_call(self, metric_fn, call_pos_args, call_kwargs):
+        """
+        Given a metric api call (include the called metric function, and call arguments)
+        Register the call information (arguments dict) into the `metric_api_call_arg_dict_list_map`
+        and return a tuple of (metric key, run_id):
+        - metric key format is: {metric_name}_on_{eval_dataset_name}
+          metric_name is generated by metric function name, if there're multiple calls with
+          different call arguments, then appending index to the name.
+        - run_id is the id of mlflow run which we need log metric into.
+        """
+        call_arg_list = list(call_pos_args) + list(call_kwargs.values())
+
+        dataset_id_list = self.pred_result_id_to_dataset_name_and_model_id.keys()
+        print(f'DGB: get_eval_dataset_name: # 1')
+
+        dataset_name = None
+        model_id = None
+
+        # Note: some metric API the arguments is not like `y_true`, `y_pred`
+        #  e.g.
+        #    https://scikit-learn.org/stable/modules/generated/sklearn.metrics.roc_auc_score.html#sklearn.metrics.roc_auc_score
+        #    https://scikit-learn.org/stable/modules/generated/sklearn.metrics.silhouette_score.html#sklearn.metrics.silhouette_score
+        for arg in call_arg_list:
+            if arg is not None and not np.isscalar(arg) and id(arg) in dataset_id_list:
+                print(f'DGB: get_eval_dataset_name: # 3')
+                dataset_name, model_id = self.pred_result_id_to_dataset_name_and_model_id[id(arg)]
+                break
+
+        if dataset_name is None or model_id is None:
+            return None
+
+        run_id = self.model_id_to_run_id_map.get(model_id, None)
+        if run_id is None:
+            return None
+
+        call_arg_dict = _extract_metric_api_call_arg_dict(metric_fn, call_pos_args, call_kwargs)
+        metric_name = metric_fn.__name__
+
+        registered_call_arg_dict_list = self.metric_api_call_arg_dict_list_map[metric_name]
+
+        index = len(registered_call_arg_dict_list)
+        for i, call_arg_dict_i in enumerate(registered_call_arg_dict_list):
+            if call_arg_dict == call_arg_dict_i:
+                index = i
+                break
+
+        if index == len(registered_call_arg_dict_list):
+            registered_call_arg_dict_list.append(call_arg_dict)
+
+        metric_key = f'{self.gen_name_with_index(metric_name, index)}_on_{dataset_name}'
```

Comment:
Just to confirm, this 

```python
sklearn.metrics.top_k_accuracy_score(y_true, y_score, k=1)
sklearn.metrics.top_k_accuracy_score(y_true, y_score, k=2)
sklearn.metrics.top_k_accuracy_score(y_true, y_score, k=3)
```

This create 3 metric, `top_k_accuracy_score_on_...`, `top_k_accuracy_score-1_on_...`, and `top_k_accuracy_score-2_on_...`

Reply:
yes.

Reply:
and each `top_k_accuracy_score` call's metric argument dict (exclude dataset argument) will be logged into "metric_info" artifact.